File: app/src/main/java/com/kickstarter/ui/activities/WebViewActivity.java
package com.kickstarter.ui.activities;

import android.os.Bundle;
import android.util.Pair;
import android.view.View;

import com.kickstarter.R;
import com.kickstarter.libs.BaseActivity;
import com.kickstarter.libs.qualifiers.RequiresActivityViewModel;
import com.kickstarter.libs.utils.AnimationUtils;
import com.kickstarter.services.KSWebViewClient;
import com.kickstarter.ui.toolbars.KSToolbar;
import com.kickstarter.ui.views.KSWebView;
import com.kickstarter.viewmodels.WebViewViewModel;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import butterknife.Bind;
import butterknife.ButterKnife;
import rx.android.schedulers.AndroidSchedulers;

import static com.kickstarter.libs.utils.TransitionUtils.slideInFromLeft;

@RequiresActivityViewModel(WebViewViewModel.ViewModel.class)
public final class WebViewActivity extends BaseActivity<WebViewViewModel.ViewModel> implements KSWebViewClient.Delegate {
  protected @Bind(R.id.web_view_toolbar) KSToolbar toolbar;
  protected @Bind(R.id.web_view) KSWebView webView;
  protected @Bind(R.id.loading_indicator_view) View loadingIndicatorView;

  @Override
  protected void onCreate(final @Nullable Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    setContentView(R.layout.web_view_layout);
    ButterKnife.bind(this);

    this.webView.client().setDelegate(this);

    this.viewModel.outputs.toolbarTitle()
      .compose(bindToLifecycle())
      .observeOn(AndroidSchedulers.mainThread())
      .subscribe(this.toolbar::setTitle);

    this.viewModel.outputs.url()
      .compose(bindToLifecycle())
      .observeOn(AndroidSchedulers.mainThread())
      .subscribe(this.webView::loadUrl);
  }

  @Override
  public void back() {
    // This logic is sound only for web view activities without RequestHandlers.
    // TODO: Refactor the client to update web history properly for activities with RequestHandlers.
    if (this.webView.canGoBack()) {
      this.webView.goBack();
    } else {
      super.back();
    }
  }

  @Override
  public void webViewExternalLinkActivated(final @NonNull KSWebViewClient webViewClient, final @NonNull String url) {}

  @Override
  public void webViewOnPageStarted(final @NonNull KSWebViewClient webViewClient, final @Nullable String url) {
    this.loadingIndicatorView.startAnimation(AnimationUtils.INSTANCE.appearAnimation());
  }

  @Override
  public void webViewOnPageFinished(final @NonNull KSWebViewClient webViewClient, final @Nullable String url) {
    this.loadingIndicatorView.startAnimation(AnimationUtils.INSTANCE.disappearAnimation());
  }

  @Override
  public void webViewPageIntercepted(final @NonNull KSWebViewClient webViewClient, final @Nullable String url) {}

  @Override
  protected @NonNull Pair<Integer, Integer> exitTransition() {
    return slideInFromLeft();
  }
}

File: app/src/main/java/com/kickstarter/ui/activities/WebViewActivity.kt
package com.kickstarter.ui.activities

import android.content.Intent
import android.os.Bundle
import androidx.activity.ComponentActivity
import androidx.activity.addCallback
import com.kickstarter.databinding.WebViewLayoutBinding
import com.kickstarter.libs.ActivityRequestCodes
import com.kickstarter.libs.Environment
import com.kickstarter.libs.utils.KsOptional
import com.kickstarter.libs.utils.extensions.addToDisposable
import com.kickstarter.libs.utils.extensions.getEnvironment
import com.kickstarter.models.User
import com.kickstarter.ui.IntentKey
import com.kickstarter.ui.data.LoginReason
import com.kickstarter.ui.extensions.finishWithAnimation
import com.kickstarter.ui.views.KSWebView
import com.kickstarter.utils.WindowInsetsUtil
import io.reactivex.Observable
import io.reactivex.disposables.CompositeDisposable

class WebViewActivity : ComponentActivity() {
    private lateinit var binding: WebViewLayoutBinding
    private lateinit var environment: Environment
    private lateinit var currentUser: Observable<KsOptional<User>>
    private val disposables = CompositeDisposable()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        this.getEnvironment()?.let { env ->
            environment = env
        }
        currentUser = requireNotNull(environment.currentUserV2()).observable()

        binding = WebViewLayoutBinding.inflate(layoutInflater)
        WindowInsetsUtil.manageEdgeToEdge(
            window,
            binding.root
        )
        setContentView(binding.root)

        val toolbarTitle = intent.getStringExtra(IntentKey.TOOLBAR_TITLE)
        toolbarTitle?.let { binding.webViewToolbar.webViewToolbar.setTitle(it) }

        binding.webView.setDelegate(object : KSWebView.Delegate {
            override fun externalLinkActivated(url: String) {
            }

            override fun pageIntercepted(url: String) {
                if (url.contains("authenticate")) {
                    finish()
                }
            }

            override fun onReceivedError(url: String) {
            }
        })

        onBackPressedDispatcher.addCallback {
            if (binding.webView.canGoBack()) {
                binding.webView.goBack()
            } else {
                finishWithAnimation()
            }
        }
        observeLoginState()
    }


    // Check if the user is logged in.
    // If no, start the LoginToutActivity. Wait for its result within the same backstack.
    // Once LoginToutActivity is finished successfully WebViewActivity can proceed to load the url to the webview.
    private fun observeLoginState() {
        currentUser
            .subscribe {
                when (it.getValue()) {
                    null -> startLoginToutActivity()
                    else -> loadUrl()
                }
            }.addToDisposable(disposables)
    }

    private fun loadUrl() {
        val url = intent.getStringExtra(IntentKey.URL)
        url?.let { binding.webView.loadUrl(it) }
    }

    private fun startLoginToutActivity() {
        val intent = Intent(this, LoginToutActivity::class.java)
            .putExtra(IntentKey.LOGIN_REASON, LoginReason.STAR_PROJECT)
        startActivityForResult(intent, ActivityRequestCodes.LOGIN_FLOW)
    }
}
